Adding customization and run & special attack bars to statusbars

Malfuryent opened the pull request:
As requested in issue #7728 

Added some functionality:
- Run Energy bar
- Special Attack bar
- The bars now stay visible with bank interface open

Added some Customization options:
- Choose which bar you want to have where or disable them selectively.

Redesigned the way the statusbars plugin works with help of Rheon.
- Moved render methods into BarRenderer
- Redesign of the Overlay

Makes use of the Poison plugin heart images and replaced the heart sprite with the one from the minimap orb.

New status bars options, you can also disable the left or right bar as you wish.
![](https://i.imgur.com/gyYvMeu.png)

Run energy & Special attack bars
![](https://i.imgur.com/l9QOafv.png)

Heart change for different status(Poison, Venom, Disease).
![](https://i.imgur.com/q7rhdoo.png) ![](https://i.imgur.com/7QeOmr3.png)![](https://i.imgur.com/wpavSnR.png) ![](https://i.imgur.com/f955QNd.png)

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/Config/BarConfig.java:
```diff
@@ -0,0 +1,48 @@
+	DISABLED("Disabled"),
+	HITPOINTS("Hitpoints"),
+	PRAYER("Prayer"),
+	RUN_ENERGY("Run_Energy"),
```
Don't have underscores in the option text

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsConfig.java:
```diff
@@ -27,13 +27,15 @@
 		keyName = "enableCounter",
-		name = "Show hitpoints & prayer counter",
+		name = "Show counters",
 		description = "Shows current amount of hitpoints & prayer on the status bars"
```
Update the description to reflect the change too

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsConfig.java:
```diff
@@ -53,11 +55,31 @@ default boolean enableSkillIcon()
 		keyName = "enableRestorationBars",
-		name = "Show amount of hitpoints and prayer restored",
+		name = "Show restores",
 		description = "Visually shows how much a food or prayer will heal/restore you on the bars."
```
Update this description too

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsConfig.java:
```diff
@@ -27,13 +27,15 @@
 import net.runelite.client.config.Config;
 import net.runelite.client.config.ConfigGroup;
 import net.runelite.client.config.ConfigItem;
+import net.runelite.client.plugins.gpu.config.AntiAliasingMode;
```
remove this

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,53 +136,168 @@ public Dimension render(Graphics2D g)
+		String leftBarText = "";
+		String rightBarText = "";
+
+		if (leftBar == "Hitpoints" || rightBar == "Hitpoints")
```
Compare against the enum, not the string.

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,53 +136,168 @@ public Dimension render(Graphics2D g)
 		}
-		else if (poisonState >= 1000000)
+
+		if (leftBar == "Prayer" || rightBar == "Prayer")
```
Compare against the enum not the string

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,53 +136,168 @@ public Dimension render(Graphics2D g)
 		}
-		else
+
+		if (leftBar == "Special_Attack" || rightBar == "Special_Attack")
```
compare against the enum not the string

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,53 +136,168 @@ public Dimension render(Graphics2D g)
-		final int currentPrayer = client.getBoostedSkillLevel(Skill.PRAYER);
-		final int quickPrayerState = client.getVar(Varbits.QUICK_PRAYER);
-		final Color prayerBar = quickPrayerState == 1 ? QUICK_PRAYER_COLOR : PRAYER_COLOR;
+		if (leftBar == "Run_Energy" || rightBar == "Run_Energy")
```
compare against the enum not the string

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,53 +136,168 @@ public Dimension render(Graphics2D g)
+			{
+				maxRightStat = 100;
+				currentRightStat = client.getEnergy();
+				rightStatusBar = new Color(199, 174, 0, 220);
```
store the colors

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,53 +136,168 @@ public Dimension render(Graphics2D g)
+			{
+				maxLeftStat = 100;
+				currentLeftStat = client.getEnergy();
+				leftStatusBar = new Color(199, 174, 0, 220);
```
store the colors

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,53 +136,168 @@ public Dimension render(Graphics2D g)
+			{
+				maxRightStat = 1000;
+				currentRightStat = client.getVar(VarPlayer.SPECIAL_ATTACK_PERCENT);
+				rightStatusBar = new Color(3, 153, 0, 195);
```
store colors

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,53 +136,168 @@ public Dimension render(Graphics2D g)
+			{
+				maxLeftStat = 1000;
+				currentLeftStat = client.getVar(VarPlayer.SPECIAL_ATTACK_PERCENT);
+				leftStatusBar = new Color(3, 153, 0, 195);
```
store colors

Malfuryent commented:
This PR quite a redesign of the original, with thanks to @Rheon-D.
- Moved render methods into BarRenderer
- Redesign of the Overlay

everything should be more readable now and less crowded.
All original functionality still works + the new functionality added like described
in the PR post.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+@RequiredArgsConstructor
+class BarRenderer
+{
+	private static final TextComponent textComponent = new TextComponent();
```
Naming scheme should be `TEXT_COMPONENT` as it's a static final variable.

raiyni replied:
why does this exist

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -27,77 +27,149 @@
+	private Image runEnergyImage;
+	private Image specialImage;
+
+	private final Map<BarMode, BarRenderer> barRenderers = new HashMap<>();
```
Use an EnumMap here.

```suggestion
	private final Map<BarMode, BarRenderer> barRenderers = new EnumMap<>(BarMode.class);
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2018, Rheon <https://github.com/Rheon-D>
```
This should be (c) 2019 and using your info, unless this was in fact written by RheonD

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		return (int) Math.round(ratio * size);
+	}
+
+	public void draw(Graphics2D graphics, int x, int y, int height)
```
Package-private
```suggestion
	void draw(Graphics2D graphics, int x, int y, int height)
```

Nightfirecat replied:
Also, what do `x` and `y` stand for? Bottom-left corner of the bar? This method might benefit from javadoc, since the parameters of the other drawing methods called from here also depend on these values.

Malfuryent replied:
x, y is the location on the screen where the bar has to be drawn, starting from the bottom left iirc yes.

How do i add javadoc? or do you mean to /* comment something in the code like this? */

Nightfirecat replied:
If you're using intellij, type `/**` and hit enter, and it should add most of the sections you'll need. I'd refer to other places in code or [the Javadoc documentation](https://www.oracle.com/technetwork/articles/java/index-137868.html) on it should be written.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -123,214 +195,86 @@ public Dimension render(Graphics2D g)
-						{
-							break;
-						}
+					if (c.getStat().getName().equals(skill))
```
Why isn't this a `continue` guard clause at the top of the loop? This can be the first thing checked before starting to parse `c.getTheoretical()`.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+	{
+		refreshSkills();
+		renderBar(graphics, x, y, height);
+		if (config.enableRestorationBars())
```
Both of these config checks seem like they should be done within `renderBar()`, as they're effectively tied to the bar rendering itself.

Also, we generally discourage brace-less statements; change this to be in this format:

```
if (config.enableRestorationBars())
{
    renderRestore(graphics, x, y, height);
}
```

Nightfirecat replied:
Also, I'd be tempted to move these config checks to within these methods since the passed parameters don't depend on the config values.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		if (config.enableSkillIcon() || config.enableCounter())
+			renderIconsAndCounters(graphics, x, y);
+	}
+}
```
Newline at EOF is preferred.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/Config/BarMode.java:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2018, Lotto <https://github.com/devLotto>
```
Same as before, use 2019 and your information.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		if (config.enableCounter())
+		{
+			graphics.setFont(FontManager.getRunescapeSmallFont());
+			textComponent.setColor(Color.WHITE);
```
It looks like `TEXT_COMPONENT` is only ever set to be white; this should be moved into the constructor instead of being called every time this method is called while `config.enableCounter()` is true.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		final int filledHeight = getBarHeight(maxValue, currentValue, height);
+		graphics.setColor(fill);
+		graphics.fillRect(x + PADDING,
+				y + PADDING + (height - filledHeight),
```
Just use a single indent here.
```suggestion
			y + PADDING + (height - filledHeight),
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		graphics.setColor(fill);
+		graphics.fillRect(x + PADDING,
+				y + PADDING + (height - filledHeight),
+				WIDTH - PADDING * OFFSET,
```
```suggestion
			WIDTH - PADDING * OFFSET,
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		graphics.fillRect(x + PADDING,
+				y + PADDING + (height - filledHeight),
+				WIDTH - PADDING * OFFSET,
+				filledHeight - PADDING * OFFSET);
```
```suggestion
			filledHeight - PADDING * OFFSET);
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+			filledHeight = filledHeight - overHeal + OVERHEAL_OFFSET;
+			graphics.setColor(OVERHEAL_COLOR);
+			graphics.fillRect(x + PADDING,
+					y - filledCurrentHeight + (height - filledHeight) + HEAL_OFFSET,
```
Single indent
```suggestion
				y - filledCurrentHeight + (height - filledHeight) + HEAL_OFFSET,
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+			graphics.setColor(OVERHEAL_COLOR);
+			graphics.fillRect(x + PADDING,
+					y - filledCurrentHeight + (height - filledHeight) + HEAL_OFFSET,
+					WIDTH - PADDING * OVERHEAL_OFFSET,
```
```suggestion
				WIDTH - PADDING * OVERHEAL_OFFSET,
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+			graphics.fillRect(x + PADDING,
+					y - filledCurrentHeight + (height - filledHeight) + HEAL_OFFSET,
+					WIDTH - PADDING * OVERHEAL_OFFSET,
+					filledHeight - PADDING * OVERHEAL_OFFSET);
```
```suggestion
				filledHeight - PADDING * OVERHEAL_OFFSET);
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		else
+		{
+			graphics.fillRect(x + PADDING,
+					y - OVERHEAL_OFFSET - filledCurrentHeight + (height - filledHeight) + HEAL_OFFSET,
```
Single indent
```suggestion
				y - OVERHEAL_OFFSET - filledCurrentHeight + (height - filledHeight) + HEAL_OFFSET,
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		{
+			graphics.fillRect(x + PADDING,
+					y - OVERHEAL_OFFSET - filledCurrentHeight + (height - filledHeight) + HEAL_OFFSET,
+					WIDTH - PADDING * OVERHEAL_OFFSET,
```
```suggestion
				WIDTH - PADDING * OVERHEAL_OFFSET,
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+			graphics.fillRect(x + PADDING,
+					y - OVERHEAL_OFFSET - filledCurrentHeight + (height - filledHeight) + HEAL_OFFSET,
+					WIDTH - PADDING * OVERHEAL_OFFSET,
+					filledHeight + OVERHEAL_OFFSET - PADDING * OVERHEAL_OFFSET);
```
```suggestion
				filledHeight + OVERHEAL_OFFSET - PADDING * OVERHEAL_OFFSET);
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+		int filledHeight = getBarHeight(maxValue, heal, height);
+		graphics.setColor(color);
+
+		if (filledHeight + filledCurrentHeight > height)
```
Does this branch of the `if` statement only draw the restoration bar? It seems like overheal should be drawn so that both current -> maximum and minimum -> final overheal value are drawn on.

That is, getting the first of these images, not the second. (run this only if the overheal exceeds maximum, but run the `else` branch of code regardless)

![full-overheal](https://user-images.githubusercontent.com/2199511/52457869-ea9dc980-2b53-11e9-9155-16a7c4a1d11f.png) ![only-overheal](https://user-images.githubusercontent.com/2199511/52457875-f6898b80-2b53-11e9-89e9-3c2ac40cb69d.png)

Malfuryent replied:
This branch only draws the restorationbar yes, i'm not entirely sure what you mean with that both should be drawn on? 
It just checks if its more than the maximum value, and if it is then it draws to full hp bar + color change do indicate overhealing which should be fine?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -123,214 +195,86 @@ public Dimension render(Graphics2D g)
-		final int currentPrayer = client.getBoostedSkillLevel(Skill.PRAYER);
-		final int quickPrayerState = client.getVar(Varbits.QUICK_PRAYER);
-		final Color prayerBar = quickPrayerState == 1 ? QUICK_PRAYER_COLOR : PRAYER_COLOR;
+		if (left != null)
```
Don't use one-line ifs here:

```
if (left != null)
{
    left.draw(g, offsetLeftBarX, offsetLeftBarY, height);
}
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+import static net.runelite.client.plugins.statusbars.StatusBarsOverlay.WIDTH;
+import net.runelite.client.ui.FontManager;
+import net.runelite.client.ui.overlay.components.TextComponent;
+
```
What's the empty line doing in the imports?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,170 @@
+			graphics.setFont(FontManager.getRunescapeSmallFont());
+			textComponent.setColor(Color.WHITE);
+			textComponent.setText(counterText);
+			textComponent.setPosition(new java.awt.Point(x + centerText + 1, y + COUNTER_ICON_HEIGHT));
```
Import `java.awt.Point` so this can just be `new Point`.

Malfuryent commented:
Resolved the requested changes, also added an offset for the counters so they are centered again.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	/**
+	 * renderBar is called to render the bar, restorationbars, icons and counters on the gamescreen
```
This should be written using indicative mood.
```suggestion
	 * Renders a status bar along with its restoration bar(s), icons and counters as appropriate
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,183 @@
+	/**
+	 * renderBar is called to render the bar, restorationbars, icons and counters on the gamescreen
+	 * @param graphics Graphics.
+	 * @param x The location on the client where it will draw the bar on the x axis.
```
This should mention which corner this is used for. I believe you said this is the bottom-left corner?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,183 @@
+	 * @param x The location on the client where it will draw the bar on the x axis.
+	 * @param y The location on the client where it will draw the bar on the y axis.
+	 * @param height The height of the bar.
+	 * @param side Is used for the Left counter offset so it draws it correctly in the center.
```
What values are accepted for this parameter? `int` doesn't really make sense here given the comment.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,181 @@
+	private final Supplier<Image> iconSupplier;
+	private int maxValue;
+	private int currentValue;
+	private final static int LEFT = 1;
```
This is no longer used.

Malfuryent commented:
Changed the behaviour of the prayer bar. It now lights up if ANY prayer is active rather than just for quickprayers. Also closes PR #6608 as that PR is old.

Unless this PR needs any other changes or requested additions it should be good to go.

Malfuryent commented:
Added a new Viewport in the latest commit for the bank interface in fixed mode, they had to be disabled previously but now that isn't necessary anymore.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/BarRenderer.java:
```diff
@@ -0,0 +1,180 @@
+		{
+			if (leftSide)
+			{
+				offset = 2;
```
Why is this offset one pixel greater when `leftSide` is true? Shouldn't the offset be the same regardless of which side the bar was drawn on?

Malfuryent replied:
This is fixed now, it now correctly draws the counters in the center for both bars.

Malfuryent commented:
Updated the PR post to reflect the changes the plugin makes. 
I also added the new poison plugin hearts to the statusbars so the sprite also changes depending on your status.

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,7 +131,27 @@ else if (client.getVar(VarPlayer.DISEASE_VALUE) > 0)
+					{
+						return HEART_POISON;
+					}
+					else if (poisonState >= 1000000)
```
the elses aren't needed

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -131,7 +131,27 @@ else if (client.getVar(VarPlayer.DISEASE_VALUE) > 0)
+					}
+					else
+					{
+						return ImageUtil.resizeCanvas(Objects.requireNonNull(spriteManager.getSprite(1067, 0)), 26, 26);
```
save this

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -174,7 +194,7 @@ else if (client.getVar(VarPlayer.DISEASE_VALUE) > 0)
 				},
 				() -> ENERGY_HEAL_COLOR,
-				() -> runEnergyImage
+				() -> ImageUtil.resizeCanvas(Objects.requireNonNull(spriteManager.getSprite(SpriteID.MINIMAP_ORB_WALK_ICON, 0)), 26, 26)
```
save this

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -184,7 +204,7 @@ else if (client.getVar(VarPlayer.DISEASE_VALUE) > 0)
 				() -> SPECIAL_ATTACK_COLOR,
 				() -> SPECIAL_ATTACK_COLOR,
-				() -> specialImage
+				() -> ImageUtil.resizeCanvas(Objects.requireNonNull(spriteManager.getSprite(1610, 0)), 26, 26)
```
save this

Malfuryent commented:
This PR is ready unless other changes or features need to be changed/added.

Malfuryent commented:
If anyone wants to take alook at this or test this, i've updated the branch to the latest snapshot. i think its ready to fit it in an update whenever you guys want. If not then i'll make changes if any are needed.

Malfuryent commented:
Updated to the latest runelite version in case someone wants to test it / look at it.

Nightfirecat commented:
Looks like it still needs rebasing since it still has conflicts in the overlay file.

Malfuryent commented:
Should be fixed now, i'm not getting any conflicts anymore.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsOverlay.java:
```diff
@@ -121,202 +228,84 @@ public Dimension render(Graphics2D g)
-	}
+			if (change != null)
+			{
+				final StatsChanges statsChanges = change.calculate(client);
```
Since this is not reused, move the call into the loop below to remove the unneeded variable definition:

```java
for (final StatChange c : change.calculate(client).getStatChanges())
```

Yttrium-tYcLief commented:
Hey @Malfuryent, any possibility of getting this wrapped up? Looks like it should be just about ready to go.

Nightfirecat commented:
I've split this into some separate commits and reworked this somewhat, but I'm of the opinion the code needs some more work before it's in a mergeable state. At the very least, this PR should not be introducing cross-plugin image references to the poison plugin's heart icon images. I plan to work on this a little further until I'm satisfied with it.

Nightfirecat commented:
@Malfuryent I've finished rebasing this into what I'd consider an acceptable PR to merge. I noticed, however, that the HP overheal overlay does not seem to work anymore after this PR is applied.

Malfuryent commented:
@Nightfirecat i updated the code with your changes but i didnt notice the problem you had, if the food i select heals to much it shows me the overheal color. Maybe you had the option to show the restores turned off in the plugin settings?

Nightfirecat commented:
Ok, the behavior I thought existed already has simply never existed. I had been under the impression that stats boosted above their maximum (through having used a brew, anglerfish, ancient mace, etc.) would display a more vivid/bright color at the bottom with the amount overhealed--say, roughly 1/5 of the bar when overhealing with an anglerfish. This has actually never been the case so it's just me being confused.

Malfuryent commented:
yeah that's right. that behaviour never existed 👍 everything else should be fine right?
